jakarta-eclipse/src/org/jakarta/jdt/diagnostics/FilterDiagnosticsCollector.java

```diff
@@ -83,15 +94,17 @@ public void collectDiagnostics(ICompilationUnit unit, List<Diagnostic> diagnosti
 						Range annotationrange = JDTUtils.toRange(unit, annotationNameRange.getOffset(), annotationNameRange.getLength());
 						
 						if (!isUrlpatternSpecified && !isValueSpecified && !isServletNamesSpecified) {
-							diagnostics.add(new Diagnostic(annotationrange, "The 'urlPatterns' attribute, 'servletNames' attribute or the 'value' attribute of the WebFilter annotation MUST be specified."));
+							diagnostic = new Diagnostic(annotationrange, "The 'urlPatterns' attribute, 'servletNames' attribute or the 'value' attribute of the WebFilter annotation MUST be specified.");
```

Comment:
Diagnostics should each have a unique code, specific to that diagnostic. If all the diagnostics are from the same specification, in this case servlet, they can have the same source

Reply:
Sorry, I missed this comment before. Now I have updated the PR. Different collectors are using different `CODE` strings now.